src/realm/util/file_mapper.cpp

```diff
@@ -104,19 +106,212 @@ struct mapping_and_addr {
+    }
+    ++info.current_version;
+}
+
+void encryption_note_reader_end(SharedFileInfo& info, void* reader_id)
+{
+    UniqueLock lock(mapping_mutex);
+    for (auto j = info.readers.begin(); j != info.readers.end(); ++j)
+        if (j->reader_ID == reader_id) {
+            // move last over
+            *j = info.readers.back();
+            info.readers.pop_back();
+            return;
+        }
+}
+
+size_t collect_total_workload() // must be called under lock
+{
+    size_t total = 0;
+    for (auto i = mappings_by_file.begin(); i != mappings_by_file.end(); ++i) {
+        SharedFileInfo& info = *i->info;
+        info.num_decrypted_pages = 0;
+        for (auto it = info.mappings.begin(); it != info.mappings.end(); ++it) {
+            info.num_decrypted_pages += (*it)->collect_decryption_count();
+        }
+        total += info.num_decrypted_pages;
+    }
+    return total;
+}
+
+/* Compute the amount of work allowed in an attempt to reclaim pages.
+ * please refer to EncryptedFileMapping::reclaim_untouched() for more details.
+ *
+ * The function starts slowly when the load is 0.5 of target, then turns
+ * up the volume as the load nears 1.0 - where it sets a work limit of 10%.
+ * Since the work is expressed (roughly) in terms of pages released, this means
+ * that about 10 runs has to take place to reclaim all pages possible - though
+ * if successful the load will rapidly decrease, turning down the work limit.
+ */
+
+struct work_limit_desc {
+    float base;
+    float effort;
+};
+const std::vector<work_limit_desc> control_table = {{0.5f, 0.001f},  {0.75f, 0.002f}, {0.8f, 0.003f},
+                                                    {0.85f, 0.005f}, {0.9f, 0.01f},   {0.95f, 0.03f},
+                                                    {1.0f, 0.1f},    {1.5f, 0.2f},    {2.0f, 0.3f}};
+
+size_t get_work_limit(size_t decrypted_pages, size_t target)
+{
+    float load = 1.0f * decrypted_pages / target;
+    float akku = 0.0f;
+    for (const auto& e : control_table) {
+    	if (load > e.base)
+    		akku += (load - e.base) * e.effort;
+    }
+    size_t work_limit = size_t(target * akku);
+    return work_limit;
+}
+
+/* Find the oldest version that is still of interest to somebody */
+uint64_t get_oldest_version(SharedFileInfo& info) // must be called under lock
+{
+    if (info.readers.size() == 0) {
+        return info.current_version;
+    }
+    else {
+        auto oldest_version = info.current_version;
+        for (const auto& e : info.readers) {
+        	if (e.version < oldest_version) {
+        		oldest_version = e.version;
+        	}
+        }
+        return oldest_version;
+    }
+}
+
+// Reclaim pages for ONE file, limited by a given work limit.
+// return true if "done for now"
+bool reclaim_pages_for_file(SharedFileInfo& info, size_t& work_limit)
+{
+    uint64_t oldest_version = get_oldest_version(info);
+    if (info.readers.size() == 0 || info.last_scanned_version < oldest_version) {
```

Comment:
Check on `info.readers.size() == 0` is redundant. If no readers, oldest_version will be equal to current_version which is always bigger than last_scanned_version.